util/timer/src/timer.rs

```diff
@@ -0,0 +1,254 @@
+        self.at.cmp(&other.at)
+    }
+}
+
+impl std::cmp::PartialOrd for TimerSchedule {
+    fn partial_cmp(&self, other: &TimerSchedule) -> Option<std::cmp::Ordering> {
+        self.at.partial_cmp(&other.at)
+    }
+}
+
+enum TimerMessage {
+    Schedule(TimerSchedule),
+    Stop,
+}
+
+fn waiter_loop(rx: Receiver<TimerMessage>, worker_queue: Arc<WorkerQueue>) {
+    let mut heap: BinaryHeap<Reverse<_>> = BinaryHeap::new();
+    loop {
+        for msg in rx.try_iter() {
+            match msg {
+                TimerMessage::Stop => return,
+                TimerMessage::Schedule(schedule) => heap.push(Reverse(schedule)),
+            }
+        }
+
+        let wait_for = handle_timeout(&worker_queue, &mut heap);
+
+        match wait_for {
+            Some(timeout) => match rx.recv_timeout(timeout) {
+                Ok(TimerMessage::Schedule(schedule)) => heap.push(Reverse(schedule)),
+                Ok(TimerMessage::Stop) => return,
+                Err(RecvTimeoutError::Disconnected) => unreachable!(), // we don't drop timer
+                Err(RecvTimeoutError::Timeout) => continue,
+            },
+            None => match rx.recv() {
+                Ok(TimerMessage::Schedule(schedule)) => heap.push(Reverse(schedule)),
+                Ok(TimerMessage::Stop) => return,
+                Err(_) => unreachable!(), // we don't drop timer loop
+            },
+        }
+    }
+}
+
+fn handle_timeout(worker_queue: &WorkerQueue, heap: &mut BinaryHeap<Reverse<TimerSchedule>>) -> Option<Duration> {
+    loop {
+        let now = Instant::now();
+        match heap.peek() {
+            None => return None,
+            Some(Reverse(earliest)) if now < earliest.at => return Some(earliest.at - now),
+            _ => { /* lifetime prevents modifying heap from here. */ }
+        }
+        let timed_out = heap.pop().unwrap().0; // doesn't panic.
+        worker_queue.enqueue(timed_out);
+    }
+}
+
+struct Workers {
+    threads: Vec<JoinHandle<()>>,
+    queue: Arc<WorkerQueue>,
+}
+
+impl Workers {
+    fn new(size: usize, timers: Arc<RwLock<TimeoutHandlerMap>>) -> Workers {
+        let queue = Arc::new(WorkerQueue::new());
+
+        let mut threads = Vec::new();
+        for i in 0..size {
+            let queue = Arc::clone(&queue);
+            let timers = timers.clone();
+            threads.push(
+                thread::Builder::new()
+                    .name(format!("Timer worker #{}", i))
+                    .spawn(move || worker_loop(timers, queue))
+                    .unwrap(),
+            );
+        }
+
+        Workers {
+            threads,
+            queue,
+        }
+    }
+
+    fn stop(self) {
+        self.queue.stop();
+        for thread in self.threads {
+            thread.join().unwrap();
+        }
+    }
+}
+
+fn worker_loop(timers: Arc<RwLock<TimeoutHandlerMap>>, queue: Arc<WorkerQueue>) {
+    loop {
+        match queue.dequeue() {
```

Comment:
Surprisingly, the worker loop can stop even there is no Stop message.
In other words, `dequeue` function can return `None` even there is no Stop message.
It seems that you think dequeue message is safe because it waits the signal, but [this function is susceptible to spurious wakeups](https://doc.rust-lang.org/std/sync/struct.Condvar.html#method.wait).
Please see [Spurious wakeup](https://en.wikipedia.org/wiki/Spurious_wakeup) for more information.